### PR TITLE
Geometry_Engine: Add Length and area method to ellipse

### LINKED ISSUE #2636: Geometry_Engine: Missing methods for Ellipse
#### Description:

Length and Area methods missing

### PR DESCRIPTION
**Pull request by IsakNaslundBh**
### Issues addressed by this PR

Closes #2636 

Adding Area and Length calculation methods for ellipse.

Area is trivial.

Length is based on infinite series calculation. 

Reference can be found here: https://en.wikipedia.org/wiki/Ellipse#Circumference (the second option by Ivory and Bessel)
And here: https://www.mathsisfun.com/geometry/ellipse-perimeter.html (Infinite series 2)

The number of terms used in the infinite series is 10 times the ratio between a and b (long and short radius) which seem to render very good results in general.

Limited the maximum number of terms to be calculated to 100 000, which will give close to exact results for ellipses with a ratio of up to 1:10 000 to 1:20 000. Calculating that number of terms takes about a minute on my machine (as a one-off calculation) that then will run quick for subsequent ellipses.

Only required number of coefficients are computed, which means even the first length calculation for more sane ellipses (rations of less than 1:100 (not sure if 1:100 should even qualify as a normal case)), will be quick to calculate even for the first pass, with subsequent calculations being even quicker. 

Can ofc increase this hard cutoff, but that seemed like a reasonable cutoff to me and as a balance for how long the computer should be allowed to be frozen. 

### Test files

Test file here:

https://burohappold.sharepoint.com/:f:/s/BHoM/EuQzVNKKlupGhuwpr3UhdFgBcLhH1RZGAu0xZyAqwPYyhA?e=xzdrsb

Some differences between the added calculation and what rhino gives, but pretty confident that the added method is more exact.

### CODE REVIEW

**Comment by IsakNaslundBh** [comment]
@BHoMBot check serialisation

**Comment by IsakNaslundBh** [comment]
@BHoMBot check core

**Comment by IsakNaslundBh** [comment]
@BHoMBot check installer

**Comment by IsakNaslundBh** [comment]
@BHoMBot check installer

File: Geometry_Engine/Query/Length.cs [hunk 1]
```diff
@@ -51,41 +59,114 @@ public static double SquareLength(this Vector vector)
...
+            //Get out a as the long radius and b as the short radius
+            double a = Math.Max(curve.Radius1, curve.Radius2);
+            double b = Math.Min(curve.Radius1, curve.Radius2);
+
+            //Special case circle
+            if (a == b)
+                return 2 * Math.PI * a;
+
+            //Special case line
+            if (b == 0)
```
Comment (pawelbaran) [inline Geometry_Engine/Query/Length.cs:97]: Wouldn't it be worth applying tolerance here?

File: Geometry_Engine/Query/Length.cs [hunk 2]
```diff
@@ -118,6 +202,81 @@ private static double Length(this ICurve curve)
...
+                    }
+                }
+            }
+            return m_ellipseCoeficients;
+        }
+
+        /***************************************************/
+
+        private static List<double> m_ellipseCoeficients = null;
+        private static object m_cooeficientLock = new object();
```
Comment (pawelbaran) [inline Geometry_Engine/Query/Length.cs:278]: Shouldn't the values be written in m_PascalCase?

File: Geometry_Engine/Query/Length.cs [hunk 3]
```diff
@@ -118,6 +202,81 @@ private static double Length(this ICurve curve)
...
+
+        [Description("Calculates the square binomial cooefficients used in the infinite series of the ellipse length calculation.")]
+        private static List<double> SquareBinomialCoefficientsEllipse(int requiredCount)
+        {
+            //Only allow for one thread to modify the cache at once.
+            lock (m_cooeficientLock)
+            {
+                //Calcuation of coefficients only needs to be done once.
+                //Coefficients are being stored for next ellipse to be evaluated
+                if (m_ellipseCoeficients == null)
```
Comment (pawelbaran) [inline Geometry_Engine/Query/Length.cs:217]: Couldn't you declare the initial value of `m_ellipseCoeficients = new List<double>();`?

**Review by pawelbaran — CHANGES_REQUESTED** [review]
LGTM functionality wise, just a few comments on the code itself.

File: Geometry_Engine/Query/Length.cs (hunk 3 above)
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/Length.cs:217]: true, will fix. Leftover from previous iteration of the code where the coefficient list was always fully populated.

File: Geometry_Engine/Query/Length.cs [hunk 4]
```diff
@@ -51,41 +59,114 @@ public static double SquareLength(this Vector vector)
...
+        [Input("curve", "The Ellipse to calculate the length of.")]
+        [Output("length", "The length of the Ellipse.", typeof(Length))]
+        public static double Length(this Ellipse curve)
+        {
+            //Get out a as the long radius and b as the short radius
+            double a = Math.Max(curve.Radius1, curve.Radius2);
+            double b = Math.Min(curve.Radius1, curve.Radius2);
+
+            //Special case circle
+            if (a == b)
```
[the remark this answers is not in the document]
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/Length.cs:93]: potentially. A bit tricky to know what tolerance to use though to ensure this works for any scale.

The check here is mainly for performance reasons and should not impact the result

File: Geometry_Engine/Query/Length.cs (hunk 1 above)
Comment (IsakNaslundBh) [reply Geometry_Engine/Query/Length.cs:97]: Same as above. This is mainly here for performance. Guess I could check if b/a < 1e-9 or something here though

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required

**Review by pawelbaran — DISMISSED** [review]
LGTM now 👍

**Comment by IsakNaslundBh** [comment]
@BHoMBot check dataset-compliance
@BHoMBot check copyright-compliance
@BHoMBot check ready-to-merge

**Comment by IsakNaslundBh** [comment]
@BHoMBot check required
@BHoMBot check dataset-compliance
@BHoMBot check copyright-compliance

**Comment by IsakNaslundBh** [comment]
@pawelbaran I had to add nullchecks, do you mind re-approving?

**Review by pawelbaran — APPROVED** [review]

**Comment by pawelbaran** [comment]
@BHoMBot check ready-to-merge

**Pull request merged**